components/script/dom/node.rs

```diff
@@ -1116,48 +1127,64 @@ impl LayoutNodeHelpers for LayoutJS<Node> {
 // Iteration and traversal
 //
 
-pub struct NodeSiblingIterator {
+pub struct SiblingIterator<T: DerivedFrom<Node>> {
     current: Option<Root<Node>>,
+    phantom: PhantomData<T>,
 }
 
-impl Iterator for NodeSiblingIterator {
-    type Item = Root<Node>;
+impl<T> Iterator for SiblingIterator<T> where T: DerivedFrom<Node> {
+    type Item = Root<T>;
 
-    fn next(&mut self) -> Option<Root<Node>> {
+    fn next(&mut self) -> Option<Self::Item> {
         let current = match self.current.take() {
             None => return None,
             Some(current) => current,
         };
-        self.current = current.GetNextSibling();
-        Some(current)
+        while let Some(next_sibling) = current.GetNextSibling() {
```

Comment:
How about

```rust
self.current = get_next(&current);
Some(current)

fn get_next(start: &Node) -> Option<Root<T>> {
    let mut current = Root::from_ref(start);
    while let Some(next) = current.GetNextSibling() {
        if let Some(derived) = Root::downcast(next) {
            return Some(derived);
        }
        current = next;
    }
    None
}
```

Also, should be use `Root<T>` rather than `Root<Node>` for the fields>